### PR TITLE
TYP fixup overloads for reset_index

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Took me longer than I'm proud to admit, but I think I've finally understood the deal with overload and default types.

We need one overload for the default case (here, `inplace=False`), and then an extra overload of each combination of keyword arguments which _precede_ `inplace`.

So here, we need:
- one case for the defaults
- one case for when there all all arguments and `inplace=True`
- one case for when there's no `level`, and `inplace=True`
- one case for when there's no `drop`, and `inplace=True`
- one case for when there's no `level`, no `drop`, and `inplace=True`

Then, `mypy` will be able to correctly infer the return type for `reset_index` based on the value of `inplace`.

This kind of thing will help resolve the mypy issues seen in #38594, where the return type from `fillna` couldn't be determined. It also means not needing to add the ignore error from #37137

----

Some examples:

```python
# t.py

import pandas as pd

reveal_type(pd.DataFrame([1,2,3]).reset_index(level=0))
reveal_type(pd.DataFrame([1,2,3]).reset_index(inplace=False))
reveal_type(pd.DataFrame([1,2,3]).reset_index(drop=True, inplace=True))
reveal_type(pd.DataFrame([1,2,3]).reset_index(level=0, inplace=True))
reveal_type(pd.DataFrame([1,2,3]).reset_index(level=0, drop=False, inplace=True))
reveal_type(pd.DataFrame([1,2,3]).reset_index(level=0, inplace=True, col_fill='a'))
reveal_type(pd.DataFrame([1,2,3]).reset_index(level=0, col_fill='a'))
reveal_type(pd.DataFrame([1,2,3]).reset_index(inplace=True))
```

```console
$ mypy t.py 
t.py:3: note: Revealed type is 'pandas.core.frame.DataFrame'
t.py:4: note: Revealed type is 'pandas.core.frame.DataFrame'
t.py:5: note: Revealed type is 'None'
t.py:6: note: Revealed type is 'None'
t.py:7: note: Revealed type is 'None'
t.py:8: note: Revealed type is 'None'
t.py:9: note: Revealed type is 'pandas.core.frame.DataFrame'
t.py:10: note: Revealed type is 'None'
```

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
> MarcoGorelli requested a review from simonjayhawkins 15 minutes ago

watch this space!

**Comment by simonjayhawkins** [comment]
> > MarcoGorelli requested a review from simonjayhawkins 15 minutes ago
> 
> watch this space!

see #40202

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5134,22 +5132,60 @@ def reset_index(  # type: ignore[misc]
...
+        level: Optional[Union[Hashable, Sequence[Hashable]]] = ...,
+        drop: bool = ...,
+        inplace: bool = ...,
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
     ) -> Optional[DataFrame]:
+        ...
+
+    def reset_index(
+        self, level=None, drop=False, inplace=False, col_level=0, col_fill=""
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:5187]: removing the type annotations from the function means that we lose the type checking for the body of the function.

so if you reveal_type on say `level` you will get `Revealed type is 'Any'`

IIUC correctly, the type annotations on the function are used to check the function itself, while the overloads are used when checking other functions that call this one.

so both are needed.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5131,15 +5129,58 @@ def reset_index(  # type: ignore[misc]
...
+
+    @overload
+    def reset_index(
+        self,
+        *,
+        inplace: Literal[True],
+        col_level: Hashable = ...,
+        col_fill: Hashable = ...,
+    ) -> None:
+        ...
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:5173]: Is the last overload needed now? won't it fall back to the function definition if no overloads match?

Comment (MarcoGorelli) [reply pandas/core/frame.py:5173]: Pretty sure it's needed, see [here](https://mypy-play.net/?mypy=latest&python=3.9&gist=98d496455f941fed31f1f66f684d5d2e) for a toy example (uncomment the commented-out part and it'll pass)

Comment (MarcoGorelli) [reply pandas/core/frame.py:5173]: Commenting out the last overload leads to the same errors observed on `master` in the typing test from https://github.com/pandas-dev/pandas/pull/40202#issuecomment-790555390

Comment (simonjayhawkins) [reply pandas/core/frame.py:5173]: ah ok, and `Revealed type is 'Any'`. I thought that it fell back to the non-overloaded function with no matching overloads. maybe something changed in mypy or only does so under different conditions.

Comment (MarcoGorelli) [reply pandas/core/frame.py:5199]: > only does so under different conditions

If I've understood correctly, this is it - see https://github.com/pandas-dev/pandas/pull/39501#issuecomment-795130774

**Review by simonjayhawkins — APPROVED** [review]
Thanks @MarcoGorelli it would be great to have some testing in place before merging this, not because there could be anything wrong with the additions, but because susequent changes could easily break things. but we are where we are, so merge when ready.

**Comment by MarcoGorelli** [comment]
Thanks - OK, merging then as I'm not sure how long it'll take for #40202 to be ready, have opened #40349 to make sure that adding tests for this isn't forgotten

**Comment by simonjayhawkins** [comment]
> have opened #40349 to make sure that adding tests for this isn't forgotten

great. we can add the reveals from your comments here to the tests in #40202

**Comment by simonjayhawkins** [comment]
just one more thing!

these extra overloads are needed because of an issue with mypy. we have now lost that link to the issue.

although the mypy issue was initially dismissed, it appears that there is acknowledgement that the workaround can becomes impractical with many arguments. There therefore may be a fix on the mypy side in the future. We somehow need to track that issue.

**Comment by MarcoGorelli** [comment]
The (*, level, inplace) overload should probably have been (level, *, inplace), but I'll check tomorrow

**Pull request merged**